### DISCUSSION TITLE
LNK11202 unresolved externals error despite #define STB_IMAGE_IMPLEMENTATION being used

### DISCUSSION THREAD

**Discussion by kakol20** [nothings/stb#1507] [Q&A] [answered]
I am compiling using Visual Studio CMake and I keep getting the error despite the fact that I have `#define STB_IMAGE_IMPLEMENTATION` being defined in the .cpp file and **not in the header file**.

Every where I look all say that I cannot define it inside a header file but nothing has worked for me.

I have used `stb_image` before and it has worked. The only difference now is that I'm using CMake. In fact, I copied a class that was using `stb_image` from another project

Is there something I have to do in CMakeLists.txt? Am I missing something here?

**Comment by nothings** [comment]
Do this:
      #define STB_IMAGE_IMPLEMENTATION
before you #include this file in *one* C or C++ file to create the implementation.

**Reply by kakol20** [reply]
I have done exactly this and it is still not working. I know there's something wrong because I have copied a class using stb_image from another project I made and it worked there.

**Reply by nothings** [reply]
Those are the instructions. Either you're following the instructions or you're not. You're saying you're following the instructions. Either you're following the instructions or you're not. I can't tell if you're following the instructions, all I can tell is that you're saying you're following the instructions, but the project is behaving like you're not.

**Reply by kakol20** [reply]
It is already in **one** file

This is my structure

I have an Image class
The header file

``` cpp
class Image {
public:
    Image(const char* file);
    Image(const Image& other);
    ~Image();
};
```
This is the cpp file

``` cpp
#include "Image.h"
#define STB_IMAGE_IMPLEMENTATION
#define STB_IMAGE_WRITE_IMPLEMENTATION

#include "stb/stb_image.h"
#include "stb/stb_image_write.h"
```

Only **ONE** file is defining them.

**Reply by nothings** [reply]
copy paste the error message

**Reply by kakol20** [reply]
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol "public: __cdecl Image::Image(char const *)" (??0Image@@QEAA@PEBDH@Z) referenced in function main	\Image\out\build\x64-debug\Image Image\out\build\x64-debug\main.cpp.obj	1	
```
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol "public: __cdecl Image::~Image(void)" (??1Image@@QEAA@XZ) referenced in function main	\Image\out\build\x64-debug\Image 	Image\out\build\x64-debug\main.cpp.obj	1
```
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK1120	2 unresolved externals  \Image\out\build\x64-debug\Image	\Image\out\build\x64-debug\Image.exe	1	
```

**Reply by rygorous** [reply]
That has nothing to do with stb_image, that's just a link error in your code.

**Reply by kakol20** [reply]
> That has nothing to do with stb_image, that's just a link error in your code.

Well, I basically copied the Image class from another project of mine - and it works there. I have made zero modifications to the Image class itself.

I'm at the beginning of a new project and I included stb_image first. Right now, I am testing reading an image and I get this error

``` cpp
Image test("test.png");
```

**Reply by rygorous** [reply, marked as answer]
Well it looks like you're not including Image.cpp in the build at all, but I reiterate, that is a problem in your build setup and with your code and has nothing to do with stb_image in any way. We can't help you debug that.

**Reply by kakol20** [reply]
This was it. I did not realise that I had to to this:

``` cmake
add_executable (STB_Image "src/main.h" "src/main.cpp" "src/image/Image.cpp")
```

Visual Studio could recognise the macro when editing but when it came to compiling, it did not.

I included `Image.h` in `main.h` assuming that CMake would automatically add the class to the build
